Sources/NIOExtras/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,393 @@
+        // final error state
+        case failed(Error)
+        // final success state
+        case completed
+    }
+
+    private var state: State = .initialized
+
+    private let targetHost: String
+    private let targetPort: Int
+    private let headers: HTTPHeaders
+    private let deadline: NIODeadline
+    private let promise: EventLoopPromise<Void>?
+
+    /// Creates a new ``NIOHTTP1ProxyConnectHandler`` that issues a CONNECT request to a proxy server
+    /// and instructs the server to connect to `targetHost`.
+    /// - Parameters:
+    ///   - targetHost: The desired end point host
+    ///   - targetPort: The port to be used when connecting to `targetHost`
+    ///   - headers: Headers to supply to the proxy server as part of the CONNECT request
+    ///   - deadline: Deadline for the CONNECT request
+    ///   - promise: Promise with which the result of the connect operation is communicated
+    public init(targetHost: String,
+                targetPort: Int,
+                headers: HTTPHeaders,
+                deadline: NIODeadline,
+                promise: EventLoopPromise<Void>?) {
+        self.targetHost = targetHost
+        self.targetPort = targetPort
+        self.headers = headers
+        self.deadline = deadline
+        self.promise = promise
+
+        self.bufferedWrittenMessages = MarkedCircularBuffer(initialCapacity: 16) // matches CircularBuffer default
+    }
+
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        switch self.state {
+        case .initialized, .connectSent, .headReceived, .completed:
+            self.bufferedWrittenMessages.append(BufferedWrite(data: data, promise: promise))
+        case .failed(let error):
+            promise?.fail(error)
+        }
+    }
+
+    public func flush(context: ChannelHandlerContext) {
+        self.bufferedWrittenMessages.mark()
+    }
+
+    public func removeHandler(context: ChannelHandlerContext, removalToken: ChannelHandlerContext.RemovalToken) {
+        // We have been formally removed from the pipeline. We should send any buffered data we have.
+        switch self.state {
+        case .initialized, .connectSent, .headReceived, .failed:
+            self.failWithError(.noResult(), context: context)
+
+        case .completed:
+            let hadMark = self.bufferedWrittenMessages.hasMark
+            while self.bufferedWrittenMessages.hasMark {
+                // write until mark
+                let bufferedPart = self.bufferedWrittenMessages.removeFirst()
+                context.write(bufferedPart.data, promise: bufferedPart.promise)
+            }
+
+            // flush any messages up to the mark
+            if hadMark {
+                context.flush()
+            }
+
+            // write remainder
+            while let bufferedPart = self.bufferedWrittenMessages.popFirst() {
+                context.write(bufferedPart.data, promise: bufferedPart.promise)
+            }
```

Comment:
There's a fairly subtle bug available here where we can inadvertently lose a flush. If one of these calls to `context.write` triggers a flush call, that flush will be lost, and so the writes will hang.

It's probably not worth being this clever, tbh: we should unconditionally flush at the end of the unbuffered writes, even if it's a little inefficient.